### ISSUE TITLE
Unspecified fields in Input object are set to null

### ISSUE THREAD

**Issue by lelong71** [graphql-java/graphql-java#106]
If my input object has the following fields:

NameInput {
   firstName: String
   lastName: String
}
and if I call

updateNameMutation(input: {
  firstName: "John"
})

the input object I get from calling DataFetchingEnvironment.getAurgument("input") in my DataFetcher returns a map with lastName set to null. This breaks PATCH since we don't know if users explicitly set a field to null or it's filled by the code.
I expected that lastName should not be in the map since users didn't pass it. However if it has defaultValue defined, it should be in the map. This is how graphql-js currently implements

Related issue filed against graphql-js - https://github.com/graphql/graphql-js/issues/133

**Comment by danielkwinsor** [comment]
We can't yet provide an `undefined` literal as javascript does, not sure what else they do in graphql-js.
DFE provides `public boolean containsArgument(String name)`  Possibly we can retrofit `containsArgument` to return false if the argument was never specified, while leaving `getArgument` the same Map style impl.

Let's watch [[RFC] Null value](https://github.com/facebook/graphql/pull/83).

**Comment by trevor-morris** [comment]
I've seen a similar issue, though for me I only saw the problem when the input object was passed as a variable. When it was passed as a literal (as in the example above), only the fields actually present in the literal seemed to be set in the map. The relevant code seems to be in ValuesResolver where there are two methods, one for converting a literal to a map, the other for converting a variable to a map.

Here's the one for converting a literal:
````java
    private Object coerceValueAstForInputObject(GraphQLInputObjectType type, ObjectValue inputValue, Map<String, Object> variables) {
        Map<String, Object> result = new LinkedHashMap<String, Object>();

        for (ObjectField objectField : inputValue.getObjectFields()) {
            GraphQLInputObjectField inputObjectField = type.getField(objectField.getName());
            // illegal field ... no corresponding key in the schema
            if (inputObjectField == null) continue;
            Object fieldValue = coerceValueAst(inputObjectField.getType(), objectField.getValue(), variables);
            if (fieldValue == null) {
                fieldValue = inputObjectField.getDefaultValue();
            }
            result.put(objectField.getName(), fieldValue);
        }
        return result;
    }
```
Note how only the fields present in the inputValue (representing the literal) get transferred to the map. However this is the method used for converting a variable to a map
````java
    private Object coerceValueForInputObjectField(GraphQLInputObjectType inputObjectType, Map<String, Object> input) {
        Map<String, Object> result = new LinkedHashMap<String, Object>();
        for (GraphQLInputObjectField inputField : inputObjectType.getFields()) {
            Object value = coerceValue(inputField.getType(), input.get(inputField.getName()));
            result.put(inputField.getName(), value == null ? inputField.getDefaultValue() : value);

        }
        return result;
    }
````
Note how all fields in the input object type are copied, regardless of whether they are actually present in the input. So, for example, if the input map contained just "firstName" = "X" and the input type had fields "firstName" and "lastName" you'd end up with a map containing "firstName"="X" and "lastName"=null

If I'm understanding this code correctly then the fix is trivial - just skip "inputField" if input.containsKey(inputField.getName()) returns false.

I am puzzled, though, that lelong71 is seeing this problem with object literals, as I only see it with input variables (e.g. if the mutation looks like updateNameMutation(input: $input), rather than updateNameMutation(input: { firstName: "New value" })). Maybe I'm misunderstanding the original issue?

**Comment by dminkovsky** [comment]
Thank you for posting. This should be fixed with #156.
